### PR TITLE
util: impl SocketOpt for AsRawFd

### PR DESCRIPTION
**Pull request by overvenus**

### CODE REVIEW

File: src/util/sockopt.rs [hunk 1]
```diff
@@ -27,12 +27,31 @@ mod unix {
...
 
-    use mio::tcp::TcpStream;
+    use std::net::TcpStream as MioTcpStream;
+    use mio::tcp::TcpStream as StdTcpStream;
     use nix::Error as NixError;
     use nix::sys::socket;
     use nix::sys::socket::sockopt;
 
-    impl SocketOpt for TcpStream {
+    impl SocketOpt for MioTcpStream {
```
Comment (siddontang) [inline src/util/sockopt.rs]: can we support `impl for AsRawFd` traits?

**Comment by overvenus** [comment]
#862 

PTAL @BusyJay

File: src/util/sockopt.rs [hunk 2]
```diff
@@ -58,10 +57,9 @@ mod unix {
 
 #[cfg(windows)]
 mod windows {
-    use mio::tcp::TcpStream;
     use std::io::Result;
 
-    impl SocketOpt for TcpStream {
+    impl<T: AsRawFd> SocketOpt for T {
```
Comment (siddontang) [inline src/util/sockopt.rs]: Does windows has this trait?

File: src/util/sockopt.rs [hunk 3]
```diff
@@ -92,19 +90,20 @@ pub use self::windows::*;
...
-    use mio::tcp::{TcpListener, TcpStream};
+    use std::net::{SocketAddr, TcpListener as StdTcpListener, TcpStream as StdTcpStream};
+    use mio::tcp::{TcpListener as MioTcpListener, TcpStream as MioTcpStream};
 
     use super::SocketOpt;
 
     #[cfg(unix)]
     #[test]
-    fn test_sock_opt() {
+    fn test_mio_sock_opt() {
```
Comment (siddontang) [inline src/util/sockopt.rs]: I think we can combine these two tests.

File: src/util/sockopt.rs (hunk 2 above)
Comment (overvenus) [inline src/util/sockopt.rs]: No, windows has AsRawSocket instead, should we use it?

Comment (siddontang) [inline src/util/sockopt.rs]: does mio and std tcp both support it?

Comment (overvenus) [inline src/util/sockopt.rs]: Oops, mio does not support AsRawSocket, maybe I should leave windows untouch.

File: src/util/sockopt.rs [hunk 4]
```diff
@@ -27,12 +27,11 @@ mod unix {
     use std::io::{Result, Error};
     use std::os::unix::io::AsRawFd;
 
-    use mio::tcp::TcpStream;
     use nix::Error as NixError;
     use nix::sys::socket;
     use nix::sys::socket::sockopt;
 
-    impl SocketOpt for TcpStream {
+    impl<T: AsRawFd> SocketOpt for T {
```
Comment (BusyJay) [inline src/util/sockopt.rs]: Why named SocketOpt?

Comment (overvenus) [inline src/util/sockopt.rs]: I think SocketOpt is short for Socket Options, any suggestion?

Comment (siddontang) [inline src/util/sockopt.rs]: the function for handing socket is set_socket_opt or get_socket_opt, I think ok.

Comment (BusyJay) [inline src/util/sockopt.rs]: I'm confused. If this is only for socket, then why impl for AsRawFd? Can I call the function on a `File` object?

Comment (siddontang) [inline src/util/sockopt.rs]: of course you can like linux setsockopt too, but it will return an error.

File: src/util/sockopt.rs [hunk 5]
```diff
@@ -101,24 +101,46 @@ mod tests {
     fn test_sock_opt() {
         let addr = "127.0.0.1:0".parse().unwrap();
 
-        let server = TcpListener::bind(&addr).unwrap();
-        let addr = &format!("{}", server.local_addr().unwrap());
-
-        let sock = TcpStream::connect(&addr.parse().unwrap()).unwrap();
+        // mio
```
Comment (BusyJay) [inline src/util/sockopt.rs]: Why not remove the test attribute and create functions `test_mio_sock_opt` and `test_std_sock_opt` call this function instead?

File: src/util/sockopt.rs [hunk 6]
```diff
@@ -101,24 +101,46 @@ mod tests {
...
-        sock.set_send_buffer_size(8192).unwrap();
-        let s2 = sock.send_buffer_size().unwrap();
-        assert!(s2 != s1, format!("{} should not equal {}", s2, s1));
-
-        sock.set_recv_buffer_size(4096).unwrap();
-        let r1 = sock.recv_buffer_size().unwrap();
-        sock.set_recv_buffer_size(8192).unwrap();
-        let r2 = sock.recv_buffer_size().unwrap();
-        assert!(r2 != r1, format!("{} should not equal {}", r2, r1));
+        mio_sock.set_send_buffer_size(4096).unwrap();
```
Comment (siddontang) [inline src/util/sockopt.rs]: can we abstract a function to test mio and std. 

```
fn test<T: AsRawFd>(t: &T) {}
```

**Comment by siddontang** [comment]
LGTM

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**